Source/DafnyRuntime/DafnyRuntimeDafny/src/dafnyRuntime.dfy

```diff
@@ -0,0 +1,865 @@
+    const values: ImmutableArray<T>
+
+    ghost predicate Valid()
+      decreases NodeCount, 0
+      ensures Valid() ==> 0 < NodeCount
+    {
+      && values.Valid()
+      && NodeCount == 1
+    }
+
+    constructor(value: ImmutableArray<T>, isString: bool := false)
+      requires value.Valid()
+      ensures Valid()
+      ensures this.values == value
+    {
+      this.values := value;
+      this.isString := isString;
+      this.NodeCount := 1;
+    }
+
+    function Cardinality(): size_t
+      requires Valid()
+      decreases NodeCount, 1
+    {
+      values.Length()
+    }
+
+    ghost function Value(): seq<T>
+      requires Valid()
+      decreases NodeCount, 2
+      ensures |Value()| < SIZE_T_LIMIT && |Value()| as size_t == Cardinality()
+    {
+      values.values
+    }
+
+    method ToArray() returns (ret: ImmutableArray<T>)
+      requires Valid()
+      decreases NodeCount, 2
+      ensures ret.Valid()
+      ensures ret.Length() == Cardinality()
+      ensures ret.values == Value()
+    {
+      return values;
+    }
+  }
+
+  class ConcatSequence<T> extends Sequence<T> {
+    const left: Sequence<T>
+    const right: Sequence<T>
+    const length: size_t
+
+    ghost predicate Valid()
+      decreases NodeCount, 0
+      ensures Valid() ==> 0 < NodeCount
+    {
+      && NodeCount == 1 + left.NodeCount + right.NodeCount
+      && left.Valid()
+      && right.Valid()
+      && left.Cardinality() as int + right.Cardinality() as int < SIZE_T_LIMIT as int
+      && length == left.Cardinality() + right.Cardinality()
+    }
+
+    constructor(left: Sequence<T>, right: Sequence<T>)
+      requires left.Valid()
+      requires right.Valid()
+      requires left.Cardinality() as int + right.Cardinality() as int < SIZE_T_LIMIT as int
+      ensures Valid()
+      ensures Value() == left.Value() + right.Value()
+    {
+      this.left := left;
+      this.right := right;
+      this.length := left.Cardinality() + right.Cardinality();
+      this.isString := left.isString || right.isString;
+      this.NodeCount := 1 + left.NodeCount + right.NodeCount;
+    }
+
+    function Cardinality(): size_t
+      requires Valid()
+      decreases NodeCount, 1
+    {
+      length
+    }
+
+    ghost function Value(): seq<T>
+      requires Valid()
+      decreases NodeCount, 2
+      ensures |Value()| < SIZE_T_LIMIT && |Value()| as size_t == Cardinality()
+    {
+      var ret := left.Value() + right.Value();
+      assert |ret| as size_t == Cardinality();
+      ret
+    }
+
+    method ToArray() returns (ret: ImmutableArray<T>)
+      requires Valid()
+      decreases NodeCount, 2
+      ensures ret.Valid()
+      ensures ret.Length() == Cardinality()
+      ensures ret.values == Value()
+    {
+      var builder := new Vector<T>(length);
+      var stack := new Vector<Sequence<T>>(TEN_SIZE);
+      AppendOptimized(builder, this, stack);
+      ret := builder.Freeze();
+    }
+  }
+
+  // Simpler reference implementation of AppendOptimized
```

Comment:
I would love it if Dafny had some way to directly prove them equivalent. Perhaps their postconditions are strong enough that you could write a lemma ensuring that. It'd be hard to use that lemma for anything, though.

Reply:
Yup, I was a little uncomfortable leaving this here since it's technically dead code, but on the other hand I think it's really useful as precise documentation and the fact that it verifies a partial proof of equivalency.